python/dolfinx/fem/petsc.py

```diff
@@ -1118,3 +1121,339 @@ def interpolation_matrix(space0: _FunctionSpace, space1: _FunctionSpace) -> PETS
+            F, form_compiler_options=form_compiler_options, jit_options=jit_options
+        )
+        if J is None:
+            V = u.function_space
+            du = ufl.TrialFunction(V)
+            J = ufl.derivative(F, u, du)
+        self._a_prec = (
+            None
+            if P is None
+            else _create_form(
+                P, form_compiler_options=form_compiler_options, jit_options=jit_options
+            )
+        )
+        self._a = _create_form(
+            J, form_compiler_options=form_compiler_options, jit_options=jit_options
+        )
+
+        self.bcs = bcs
+        self.u = u
+
+    @property
+    def comm(self):
+        """The MPI communicator associated with the problem"""
+        return self._comm
+
+    @property
+    def L(self) -> Form:
+        """Compiled linear form (the residual form)"""
+        return self._L
+
+    @property
+    def a(self) -> Form:
+        """Compiled bilinear form (the Jacobian form)"""
+        return self._a
+
+    def update_solution(self, x: PETSc.Vec):
+        """Update the solution vector in the function used in the residual and Jacobian.
+
+        Args:
+            x: The vector containing the latest solution
+        """
+        x.ghostUpdate(addv=PETSc.InsertMode.INSERT, mode=PETSc.ScatterMode.FORWARD)
+        x.copy(self.u.x.petsc_vec)
+        self.u.x.petsc_vec.ghostUpdate(addv=PETSc.InsertMode.INSERT, mode=PETSc.ScatterMode.FORWARD)
+
+    def copy_solution(self, x: PETSc.Vec):
+        """Copy the solution vector from the function used in the residual and Jacobian.
+
+        Args:
+            x: The vector containing the latest solution
+        """
+        self.u.x.petsc_vec.copy(x)
+        x.ghostUpdate(addv=PETSc.InsertMode.INSERT, mode=PETSc.ScatterMode.FORWARD)
+
+    def F(self, snes: PETSc.SNES, x: PETSc.Vec, F: PETSc.Vec):
+        """Assemble the residual F into the vector b.
+
+        Args:
+            snes: The solver instance
+            x: The vector containing the latest solution
+            F: Vector to assemble the residual into
+        """
+        x.ghostUpdate(addv=PETSc.InsertMode.INSERT, mode=PETSc.ScatterMode.FORWARD)
+        x.copy(self.u.x.petsc_vec)
+        self.u.x.petsc_vec.ghostUpdate(addv=PETSc.InsertMode.INSERT, mode=PETSc.ScatterMode.FORWARD)
+        with F.localForm() as f_local:
+            f_local.set(0.0)
+        dolfinx.fem.petsc.assemble_vector(F, self.L)
+        dolfinx.fem.petsc.apply_lifting(F, [self.a], bcs=[self.bcs], x0=[x], alpha=-1.0)
+        F.ghostUpdate(addv=PETSc.InsertMode.ADD, mode=PETSc.ScatterMode.REVERSE)
+        dolfinx.fem.petsc.set_bc(F, self.bcs, x, -1.0)
+
+    def J(self, snes, x, J, P):
+        """Assemble the Jacobian matrix.
+
+        Args:
+            snes: The solver instance
+            x: The vector containing the latest solution
+            J: Matrix to assemble the Jacobian into
+            P: Matrix to assemble the preconditioner into
```

Comment:
What happens if the preconditioner depends on `x`? This is infact almost always the case.